### PR TITLE
Sequence-aware indexing refactor

### PR DESCRIPTION
**Pull request by tkporter**
### Description

Part 1 of an indexing refactor. This mostly focuses on the sequence-aware cursors.

Changes:
- Some renames, e.g. `SequenceIndexer` -> `SequenceAwareIndexer`, `sequence_and_tip` -> `latest_sequence_count_and_tip`
- Moved cursors out of `contract_sync/cursor.rs` into `contract_sync/cursors/*`
- The goal was to make the the Forward sequence-aware cursor, the Backward sequence-aware, and the ForwardBackward sequence aware cursors more resilient, easier to understand, and better tested.
- General strategy for doing this was to:
  - Only move onto a new range to index if the logs passed into the cursor's `update` function justify this
  - If updated with any weird data, like gaps in logs, rewind the cursor and retry previously queried ranges

I'd recommend starting the review in the following order:

1. `rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs` - where `ForwardSequenceAwareSyncCursor` & tests live
2. `rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs` - where `BackwardSequenceAwareSyncCursor` & tests live
3. `rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs` - where `ForwardBackwardSequenceAwareSyncCursor` lives

### CODE REVIEW

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs
Comment (tkporter) [inline rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs:105]: These TODOs (& the ones in the Forward & Backward cursors) were present before & I'll leave that out of scope for now

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (tkporter) [inline rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:271]: I'm realizing the range checking could be more forgiving - e.g. if the expected range is 100..=104 and we get logs with sequences [100, 101, 103, 104] (missing 102), as it's written now it'll throw out the range and rewind to try looking for 100 again. Instead it could accept 100 and 101 and then start looking for 102.

I'm open to making a change to have this be considered - however because we will have still written all the logs to the log store, and because we have logic to fast-forward, we end up having pretty much that same behavior. On the next call to `next_action`, it'll fast forward past 100 and 101 because it'll see that it's already in the db, and then start looking for 102

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (daniel-savu) [inline rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:154]: I'd add a doc comment to this function to explain that fast-forwarding is possible due to the local db, and how that's done

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (daniel-savu) [inline rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:166]: Should we rename this to `retrieve_block_number_by_sequence`?

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (daniel-savu) [inline rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:119]: At this point the indentation level is a bit high for a single function, making it hard to reason about. We should move this into another function. The u32::min range logic can be shared between `Block` and `Sequence`, which would make it even easier to understand

**Review by daniel-savu — COMMENTED** [review]
Took lots of context loading, only a few comments for now but more coming soon

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (daniel-savu) [inline rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:836]: If you pass the sequences out of order, e.g. `4`, `5`, `6`, `5`, the test will fail bc of the dedupe behavior I mentioned above

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (daniel-savu) [inline rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:931]: If I set the return value of `latest_sequence_querier` above to `sequence: 5, at_block: 88` (updating the target snapshot assertion as well), this call to `update` still works, which feels odd. Should we also filter out entries with `block > target_snapshot.block` in `update`?

**Review by daniel-savu — COMMENTED** [review]
Mostly done reviewing the forward cursor so far

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs:33]: this has different behavior when self.sequence is None, it'll set to 1 instead of 0

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:79]: Idk if it's worth spending much time on this given plans around https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3266 and how this'll boil down to a runtime check anyways

File: rust/hyperlane-base/src/contract_sync/mod.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/mod.rs:65]: Open to this but I think there's something nice about having cursors being a self contained way to figure out what action should be done and having the ContractSync::sync loop take the action

Can we leave this to https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3264 to figure out whether to do? As the watermark cursor more heavily involves the sleep action - I think it'll be more clear when doing that one whether it's more natural to change to next_range

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:119]: will change the indentation but the values in the u32::min range are different, maybe i'm misinterpreting what you mean though

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:931]: doesn't setting the latest sequence count to 5 result in a None action? or did you mean sequence: 6

imo `update` can assume that any logs passed into the cursor are legitimate, and it should succeed as long as it builds upon the last indexed snapshot and there are no gaps in the logs

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs:33]: right, I keep getting confused by the difference between sequence and count

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:79]: Hmm it's not just about the runtime check, but also making the code more readable. It's pretty confusing IMO to have the returned `RangeInclusive` mean different things when the function is called with a different `self` state. Let's do #3266 first and see afterwards

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:253]: mm this is a good point

I think my initial goal was to have internal state only relate to things in service of making sure that we're building upon the last indexed sequence, and to have `update` avoid making assumptions about how the cursor is used and to just ensure that the logs passed in build upon the last indexed sequence, but I think the delineation isn't super clear. I'll come back to this

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:119]: > values in the u32::min range are different

Fair, I didn't look closely enough

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:931]: Indeed I meant sequence 6. What I mean is - what if there's a reorg? 

 You'd end up with a `TargetSnapshot { seq: 6, block: 88 }`, and `LastIndexedSnapshot { seq: 5, block: 115 }`, but also with a `get_next_range` = `110..=88`

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:253]: Ideally the cursor implemented the `Iterator` trait but that's not really possible since it has internal state to mutate

Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:253]: After mulling it over I think having the range explicitly passed in is probably helpful. Curious what you think

The cursor update is really dependent on the logs + range, not just the logs. I think it also could reduce misuse of cursors in the future - by putting a range in the function signature it makes it clear that the cursor updates rely heavily on the range. I think part of the messiness of the prior implementations was that it wasn't super clear what was happening inside a cursor or how a cursor should be used, I think if we put the range as internal state we now implicitly closely tie the cursor's `update` fn to the `get_action` fn

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs]: moved to unique_by just to avoid the dependence on it being sorted

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:407]: Hmm I think there's potential for misuse if the range is required too, since it's not entirely clear that the range to pass in should be the one from `get_action`. I'm ok either way, just wanted to point this out

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:1003]: mm good question

I think it's probably okay to not worry about logic to accommodate reorgs. The `latest_sequence_count_and_tip` is meant to return only the finalized sequence count & the finalized tip. So as long as we've configured the node with a correct reorg period it should never encounter a reorg, so imo we can assume that all logs that make it to the `update` function are finalized. If the agent is misconfigured then I think it's ok to have undefined behavior - but certainly adding logs in such a case to complain loudly makes sense to me, I'll add some

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:78]: sounds good, I agree it's confusing, happy to consider completion criteria for 3266 to generally make sequence ranges vs block ranges less confusing

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs:106]: tracking this here https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3265

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:702]: Good call, I'll make a pass through it

**Review by daniel-savu — COMMENTED** [review]
Done with the backward cursor as well. The unit tests tend to only cover the happy path, would be a good idea to include more duplicates and out-of-order logs in there (I already did this on my local and all works fine). We can actually check if we missed some obvious spots in the logic by having a look at the test coverage report (following [this guide](https://blog.rng0.io/how-to-do-code-coverage-in-rust/) for instance)

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs
Comment (daniel-savu) [inline rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs:122]: the doc comment of this function is `The latest block that has been queried`, rather than the latest block of the chain. I feel like always returning the forward cursor's latest block violates that, and we should either change the doc comment or the name of the method to `latest_queried_block`

File: rust/hyperlane-base/Cargo.toml
Comment (daniel-savu) [inline rust/hyperlane-base/Cargo.toml:43]: Looks like it's only used in tests, let's move under `dev-dependencies`?

File: rust/hyperlane-core/src/traits/cursor.rs
Comment (daniel-savu) [inline rust/hyperlane-core/src/traits/cursor.rs:20]: Should it also include a timestamp along with the block number? =

File: rust/hyperlane-core/src/traits/db.rs
Comment (daniel-savu) [inline rust/hyperlane-core/src/traits/db.rs:34]: Nit: unrelated to this PR, but I think this `HyperlaneSequenceIndexerStore` trait should be renamed to `SequenceIndexerStoreReader`. Initially I thought the cursors had write access to the db due to this trait name, but turns out it's just reading for fast-forwarding purposes

**Review by daniel-savu — APPROVED** [review]
LFG

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs]: agree there's room for deduping stuff between the two. I actually intentionally avoided trying to share logic because this felt like a source of confusion & bugs in the past implementation

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs]: great catch, this was a bug that is now covered by unit tests

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs:122]: agree this is confusing, changed to `latest_queried_block`. We can also spend more time on this in https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3265 - it's not clear to me that the latest queried block is the right proxy for healthy either

File: rust/hyperlane-core/src/traits/cursor.rs
Comment (tkporter) [reply rust/hyperlane-core/src/traits/cursor.rs:20]: added to https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3265

File: rust/hyperlane-core/src/traits/db.rs
Comment (tkporter) [reply rust/hyperlane-core/src/traits/db.rs:34]: +1, I'll just rename, this is confusing

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:710]: I spent a bit of time on this and felt it made the tests harder to understand, happy to change if you feel strongly tho

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs
Comment (tkporter) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs:198]: The annoying part is disambiguating between a zero sequence when we do
```
TargetSnapshot {
                sequence: current_indexing_snapshot.sequence.saturating_sub(logs_len),
```
to mean that we've indexed sequence zero already (which would happen if logs_len > current_indexing_snapshot.sequence), or to mean that we now want to look for index 0

But yeah this is a bit confusing, I'll change to something hopefully more idiomatic

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs:719]: my ideal unit test looks almost like a parameterized function, so I can play around with its input without changing many lines of code, to see if the tested condition still holds. Ok to leave as-is but that's my preference

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs]: fair

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs
Comment (daniel-savu) [reply rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs]: mm right, so we can have `current_indexing_snapshot.sequence` = 0 and still not be done with indexing. Thanks for clarifying with that comment

File: rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs
Comment (daniel-savu) [inline rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/backward.rs:189]: slick

**Review by daniel-savu — APPROVED** [review]

**Pull request merged**
